File: backend/src/main/java/com/contentplatform/backend/interfaces/web/controller/PublicPageMenuController.java
package com.contentplatform.backend.interfaces.web.controller;

import com.contentplatform.backend.application.service.PageMenuService;
import com.contentplatform.backend.domain.value.MenuLocation;
import com.contentplatform.backend.interfaces.web.response.MenuResponse;
import com.contentplatform.backend.interfaces.web.response.PageContentResponse;
import org.springframework.http.ResponseEntity;
import org.springframework.web.bind.annotation.GetMapping;
import org.springframework.web.bind.annotation.PathVariable;
import org.springframework.web.bind.annotation.RequestHeader;
import org.springframework.web.bind.annotation.RequestMapping;
import org.springframework.web.bind.annotation.RequestParam;
import org.springframework.web.bind.annotation.RestController;

import java.util.List;
import java.util.Map;

@RestController
@RequestMapping("/api/v1/content")
public class PublicPageMenuController {
    private final PageMenuService pageMenuService;

    public PublicPageMenuController(PageMenuService pageMenuService) {
        this.pageMenuService = pageMenuService;
    }

    @GetMapping("/pages")
    public ResponseEntity<List<PageContentResponse>> listPages(@RequestHeader("application-id") String applicationId,
                                                               @RequestHeader("application-token") String token,
                                                               @RequestParam(required = false) String languageCode) {
        String appId = pageMenuService.validateApplicationToken(applicationId, token);
        return ResponseEntity.ok(pageMenuService.listPublishedPages(appId, languageCode));
    }

    @GetMapping("/pages/slugs")
    public ResponseEntity<List<Map<String, String>>> listPageSlugs(@RequestHeader("application-id") String applicationId,
                                                                   @RequestHeader("application-token") String token,
                                                                   @RequestParam(required = false) String languageCode) {
        String appId = pageMenuService.validateApplicationToken(applicationId, token);
        return ResponseEntity.ok(pageMenuService.listPublishedPages(appId, languageCode).stream()
            .map(page -> Map.of("slug", page.getSlug(), "languageCode", page.getLanguageCode()))
            .toList());
    }

    @GetMapping("/pages/{languageCode}/{slug}")
    public ResponseEntity<PageContentResponse> getPage(@RequestHeader("application-id") String applicationId,
                                                       @RequestHeader("application-token") String token,
                                                       @PathVariable String languageCode,
                                                       @PathVariable String slug) {
        String appId = pageMenuService.validateApplicationToken(applicationId, token);
        return ResponseEntity.ok(pageMenuService.getPublishedPage(appId, languageCode, slug));
    }

    @GetMapping("/menus/location/{languageCode}/{location}")
    public ResponseEntity<List<MenuResponse>> getMenusByLocation(@RequestHeader("application-id") String applicationId,
                                                                 @RequestHeader("application-token") String token,
                                                                 @PathVariable String languageCode,
                                                                 @PathVariable MenuLocation location) {
        String appId = pageMenuService.validateApplicationToken(applicationId, token);
        return ResponseEntity.ok(pageMenuService.getPublicMenusByLocation(appId, languageCode, location));
    }

    @GetMapping("/menus/{languageCode}/{code}")
    public ResponseEntity<MenuResponse> getMenu(@RequestHeader("application-id") String applicationId,
                                                @RequestHeader("application-token") String token,
                                                @PathVariable String languageCode,
                                                @PathVariable String code) {
        String appId = pageMenuService.validateApplicationToken(applicationId, token);
        return ResponseEntity.ok(pageMenuService.getPublicMenuByCode(appId, languageCode, code));
    }
}

File: backend-nestjs/src/controllers/public-page-menu.controller.ts
import { Controller, Get, Param, Query, Req, UseGuards } from '@nestjs/common';
import { ApiSecurity } from '@nestjs/swagger';
import type { Request } from 'express';
import { ApplicationTokenGuard } from '../auth/application-token.guard';
import { MenuLocation } from '../common/menu-types';
import { PageContentResponseDto } from '../dto/responses/page-response.dto';
import { MenuResponseDto } from '../dto/responses/menu-response.dto';
import { ApplicationEntity } from '../entities/application.entity';
import { DomainPolicyService } from '../services/domain-policy.service';
import { AdminMenuService } from '../services/admin-menu.service';
import { AdminPageService } from '../services/admin-page.service';

@Controller('/api/v1/content')
@UseGuards(ApplicationTokenGuard)
@ApiSecurity({ 'application-id': [], 'application-token': [] })
export class PublicPageMenuController {
  constructor(
    private readonly pageService: AdminPageService,
    private readonly menuService: AdminMenuService,
    private readonly domainPolicy: DomainPolicyService,
  ) {}

  @Get('pages')
  async listPages(@Req() request: Request & { application?: ApplicationEntity }, @Query('languageCode') languageCode?: string): Promise<PageContentResponseDto[]> {
    const application = this.getApplication(request);
    this.enforceDeliveryDomainPolicy(application, request);
    return this.pageService.listPublished(application, languageCode);
  }

  @Get('pages/slugs')
  async listPageSlugs(@Req() request: Request & { application?: ApplicationEntity }, @Query('languageCode') languageCode?: string): Promise<Array<{ slug: string; languageCode: string }>> {
    const pages = await this.listPages(request, languageCode);
    return pages.map((page) => ({ slug: page.slug, languageCode: page.languageCode }));
  }

  @Get('pages/:languageCode/:slug')
  async getPage(@Req() request: Request & { application?: ApplicationEntity }, @Param('languageCode') languageCode: string, @Param('slug') slug: string): Promise<PageContentResponseDto> {
    const application = this.getApplication(request);
    this.enforceDeliveryDomainPolicy(application, request);
    return this.pageService.getPublished(application, languageCode, slug);
  }

  @Get('menus/location/:languageCode/:location')
  async getMenusByLocation(
    @Req() request: Request & { application?: ApplicationEntity },
    @Param('languageCode') languageCode: string,
    @Param('location') location: MenuLocation,
  ): Promise<MenuResponseDto[]> {
    const application = this.getApplication(request);
    this.enforceDeliveryDomainPolicy(application, request);
    return this.menuService.getPublicByLocation(application, languageCode, location);
  }

  @Get('menus/:languageCode/:code')
  async getMenu(@Req() request: Request & { application?: ApplicationEntity }, @Param('languageCode') languageCode: string, @Param('code') code: string): Promise<MenuResponseDto> {
    const application = this.getApplication(request);
    this.enforceDeliveryDomainPolicy(application, request);
    return this.menuService.getPublicByCode(application, languageCode, code);
  }

  private getApplication(request: Request & { application?: ApplicationEntity }): ApplicationEntity {
    return request.application as ApplicationEntity;
  }

  private enforceDeliveryDomainPolicy(application: ApplicationEntity, request: Request): void {
    if (!this.domainPolicy.hasOriginSignal(request)) {
      return;
    }
    this.domainPolicy.ensureAllowed(application, request);
  }
}
